### PR TITLE
Enable configuration of micrometer global registry

### PR DESCRIPTION
**Pull request by robobario**
Implements #111 

This enables configuration of some key micrometer features:
1. common tags to add to all metrics (labels in prometheus-speak)
2. binders (components that can provide common metrics like jvm or system metrics)

and adds a configurable hook so you could apply arbitrary configuration to the global registry.

You can configure common tags to add to all metrics by adding configuration like this to the kroxy yaml:

```
micrometer:
  commonTags:
    zone: euc-1a
```

You can configure binders to bind to the global registry by adding configuration like this to the kroxy yaml:

```
micrometer:
  binders:
    - JvmGcMetrics
    - io.micrometer.core.instrument.binder.jvm.JvmHeapPressureMetrics
```

First we check if there is a class that can be loaded directly using the string as a type name, otherwise we will try to load it from these packages:

* io.micrometer.core.instrument.binder.jvm
* io.micrometer.core.instrument.binder.system

Note it is assumed that the binder implementations have a no-arg constructor if you need something more complex you could use the configuration hook.

You can specify a MicrometerConfigurationHook implementation as a way to perform arbitrary configuration of the global MeterRegistry by adding configuration like this to the kroxy yaml:

```
micrometer:
  configurationHookClass: your.custom.HookClass
```

This hook will be called after the `binders` and `common tags` are configured

### CODE REVIEW

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:200]: Personally I think we should have separate test cases for each of the `default labels`, `binder` and `config hook`. 

I'd also prefer 
```suggestion
            org.assertj.core.api.Assertions.assertThat(response.body()).contains("jvm_gc_memory_allocated_bytes_total");
```

Note that deliberately doesn't include the labels so that further changes to the default set don't affect this test.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/KafkaProxy.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/KafkaProxy.java:91]: I think the `null` check and using `MicrometerConfiguration.empty()` should be pushed down into `config.micrometerConfig`

File: kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java:31]: Do we need this flag? Wouldn't callers who set it to true just do  
```
loadClass(superClass, configurationHookClass, true)
                 .orElseThrow(() -> new IllegalArgumentException("could not load " + configurationHookClass));
```

File: kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java:32]: Needs to be a more generic error message or accept it as a parameter.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java:20]: `configurationHookClass` -> `requestedClass` given this isn't specific to configuration hooks.

File: kroxylicious/src/test/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationTest.java
Comment (SamBarker) [inline kroxylicious/src/test/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationTest.java:22]: I think the tests would read better using `org.assertj.core.api.Assertions.assertThat`
e.g.

```java
    @Test
     public void testLocatesJvmBinder() {
         MicrometerConfiguration config = new MicrometerConfiguration(List.of("JvmGcMetrics"), null, null);
         List<Class<? extends MeterBinder>> binders = config.getBinders();
         assertThat(binders).contains(JvmGcMetrics.class);
     }
```

File: integrationtests/src/test/java/io/kroxylicious/proxy/KroxyConfigBuilder.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KroxyConfigBuilder.java:46]: Should we take some action in the event of duplicate tags?

**Review by SamBarker — APPROVED** [review]
LGTM

Some style questions but otherwise looks good.

**Comment by SamBarker** [comment]
@robobario and I were just chatting about this on a call and we started discussing the question of filters generating metrics and is it safe for them to assume that the global registry is available and appropriate. 

My use case is a hypothetical request response count filter which keeps a simple count of requests and responses received by the frontend. In a multi tenant world does that filter need to become tenant aware (seems dubious) or should we have a API method on the context to provide meters or the meter registry (where each registry could be configured with a tag identifying the tenant)?

@tombentley @k-wall any thoughts about how to handle this? I'm a bit worried about the need to leak the idea of tenants to all filters? Or maybe we need to expose the tenancy as first class concept in the context API so that its available to filter authors?

**Comment by tombentley** [comment]
The filter API tries to ensure that protocol filters compose nicely. A protocol filter shouldn't know about any filters before or after it in the chain. It should be able to be written as if it were the only filter in the chain, and if it's deployed with other filters the individual behaviour should still work, and you should get the combined effect of all the plugins. So it's a good point about MT. It would be good to have metrics generated from later filters in the chain be tagged with the tenant id, but without them knowing that the MT filter was before them in the chain. 

While the munging of things like topic names would help distinguish metrics from different tenants where a topic name is used in a tag, it doesn't seem reliable (not all metrics could be tagged with a topic name or consumer group id, for example) or convenient (harder to extract the tenant id for use in alerting/dashboards).

FWIW Kafka itself has some similar considerations where plugins need to expose metrics. [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics) is an attempt to address that there. Kafka has its own API for adding metrics and recording events. IDK the micrometer API at all, but I wonder if it might be worth building some mechanism for achieving the same thing. That probably involves changes to how filters are initialized and better defining the plugin lifecycle, so I don't think it should block this PR. It's better to merge this PR and consider the plugin lifecycle separately.

One other observation on this PR is that it adds more extensions point where we reflectively instantiate classes based on configuration. The more we do this the harder it is to make the proxy quarkus/GraalVM-friendly. If micrometer supports only a few binders OOTB then it might be worth supporting them explicitly (i.e. instantiate them directly, rather than reflectively) based on the short form name (e.g. `JvmGcMetrics`). And defer the decision about supporting arbitrary binders via reflection using fully qualified class names until we've had a chance to consider native compilation properly.
Relatedly, what we're doing with these extension points is slowly building a kind of ad-hoc dependency injection mechanism. Using quarkus would force us into a different path (relying on CDI annotations) which might ultimately make more sense. So it might be worth having a discussion about the merits of that, as well.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:200]: Have hit a bit of a snag on this one.

The micrometer global registry is static and has some configuration you can't clean out between tests. So the tests were non-deterministic, like if you have a separate test for common tags, then depending on test order you see the tags unexpectedly in other tests.

**Comment by SamBarker** [comment]
> That probably involves changes to how filters are initialized and better defining the plugin lifecycle, so I don't think it should block this PR. It's better to merge this PR and consider the plugin lifecycle separately.

For the record I'm 100% in agreement. This can merge without an answer to the multi tenant question

File: kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java:50]: `Optional<Class<...>>` instead of a `null` return value.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java:71]: `findFirst` rather than `findAny` so that loading is deterministic. 

I need to do some research around how reflection/class loading work in a Graal world after @tombentley mentioning it being possibly problematic for native. 

I like the idea of only offer named bindings, but that also feels quite restrictive for people deploying Kroxy they can only use the binders we have approved.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationHook.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationHook.java:11]: nit:
```suggestion
    void configure(MeterRegistry targetRegistry);
```

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/MeterRegistries.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/MeterRegistries.java:36]: nit: 
```suggestion
    private void configureCommonTags(MicrometerConfiguration micrometerConfig, CompositeMeterRegistry targetRegistry) {
```

**Comment by robobario** [comment]
With micrometer it would be possible to construct new `CompositeRegistry` instances to do the tagging. You do have to be careful because the underlying prometheus client can't handle metrics being registered with the same name but different tags.

Here's an example:
```java
import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
import io.micrometer.prometheus.PrometheusConfig;
import io.micrometer.prometheus.PrometheusMeterRegistry;

public class MicrometerTest {

    public static void main(String[] args) {
        PrometheusMeterRegistry prometheus = new PrometheusMeterRegistry(PrometheusConfig.DEFAULT);
        prometheus.counter("event_total", "a", "b").increment();
        prometheus.counter("event_total", "c", "d").increment();
        System.out.println(prometheus.scrape());
        // the bad thing, the second event_total meter registered isn't reflected in the scrape
        /*
         * # HELP event_total
         * # TYPE event_total counter
         * event_total{a="b",} 1.0
         */

        prometheus = new PrometheusMeterRegistry(PrometheusConfig.DEFAULT);
        CompositeMeterRegistry registryA = new CompositeMeterRegistry().add(prometheus);
        registryA.config().commonTags("tenant", "bank");
        CompositeMeterRegistry registryB = new CompositeMeterRegistry().add(prometheus);
        registryB.config().commonTags("tenant", "dairy");
        registryA.counter("event_total", "type", "Fetch").increment();
        registryB.counter("event_total", "type", "Produce").increment();
        System.out.println(prometheus.scrape());
        /*
         * # HELP event_total
         * # TYPE event_total counter
         * event_total{tenant="bank",type="Fetch",} 1.0
         * event_total{tenant="dairy",type="Produce",} 1.0
         */
    }
}
```

**Comment by k-wall** [comment]
lgtm, thanks @robobario

**Comment by tombentley** [comment]
Unless anyone has any more comments I guess you can merge this @robobario.

**Comment by robobario** [comment]
Thanks, I'll clean up the nits and merge

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java]: Pulled it into it's own test class with a method testing each feature, and made the assertions label agnostic. I also configured surefire and failsafe to use random test ordering while I was at it, to show up problems like that.

Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java]: Keen to go back and do further cleaning in the tests, there's a lot of duplicate creation of kafka brokers and proxies. But I figure touching that conflicts with the coming changes extracting the broker creation?

File: kroxylicious/src/main/java/io/kroxylicious/proxy/KafkaProxy.java
Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/KafkaProxy.java]: done

File: kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java
Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java]: refactored this so that it doesn't have the `failIfNotFound` flag, the error is now handled higher up where it was trying to load the hook.

File: kroxylicious/src/test/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationTest.java
Comment (robobario) [reply kroxylicious/src/test/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationTest.java:23]: switched to assertj

File: kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java
Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java]: fixed the naming in this class

File: kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java
Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/classloader/ClassloaderUtils.java]: refactored to only have a `loadClassIfPresent` method and leave it to callers to react to missing classes

File: kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java
Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java]: done. I also made the field `Optional` despite the general recommendation not to (because `Optional` is not Serializable).

File: kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java
Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfiguration.java]: Changed to `findFirst()`.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationHook.java
Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationHook.java]: fixed the naming

File: integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java:48]: Should the reference to the JVMThreadMetrics be captured and closed at tear down or similar?

File: integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java:54]: Should be pulled out as a field.

File: integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java:41]: I think this could be injected directly into each test method, but I don't know if it makes any practical difference.

File: integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java:92]: Should proxy be shutdown in a finally block?

Or maybe store a reference to it in a field and shutdown in an after all. I'd really prefer to see the proxy injected along with the Kafka cluster but that's for an other day.

File: kroxylicious/src/test/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationTest.java
Comment (SamBarker) [inline kroxylicious/src/test/java/io/kroxylicious/proxy/config/micrometer/MicrometerConfigurationTest.java:36]: Yes! That reads better and is less fragile to boot 😁

**Review by SamBarker — APPROVED** [review]

File: integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java:49]: Added a `clear()` of the globalRegistry to the beforeEach.

File: integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java]: Moved the shutdown into an `afterEach` (we need a fresh proxy every time because of the varying proxy configuration)

File: integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/MetricsIT.java]: done

**Pull request merged**
